File: Low Level Design (LLD)/DesignPatterns/3. Structural/State/Cpp/src/ModerationState.cpp
#include <iostream>
#include "../include/ModerationState.h"
#include "../include/PublishedState.h"
#include "../include/Document.h"

void ModerationState::edit(Document* doc) {
    std::cout << "Cannot edit while in moderation.\n";
}

void ModerationState::publish(Document* doc) {
    std::cout << "Publishing document.\n";
    doc->setState(new PublishedState());
}

File: Low Level Design (LLD)/DesignPatterns/3. Structural/State/Java/ModerationState.java
public class ModerationState implements State {
    @Override
    public void edit(Document doc) {
        System.out.println("Cannot edit while in moderation.");
    }

    @Override
    public void publish(Document doc) {
        System.out.println("Publishing document.");
        doc.setState(new PublishedState());
    }
}
